src/appservice/Intent.ts

```diff
@@ -270,20 +268,16 @@ export class Intent {
      */
     @timedIntentFunctionCall()
     public async ensureJoined(roomId: string) {
-        if (this.knownJoinedRooms.indexOf(roomId) !== -1) {
-            return;
+        if (!this.knownJoinedRooms) {
+            await this.refreshJoinedRooms();
         }
 
-        await this.refreshJoinedRooms();
-
-        if (this.knownJoinedRooms.indexOf(roomId) !== -1) {
+        if (this.knownJoinedRooms!.has(roomId)) {
```

Comment:
Since this is the only lookup of `knownJoinedRooms`, maybe it's worth getting rid of it after all?

The only time `ensureJoined` is called is when sending text events, which admittedly a hotspot, but what may work better than an "ensure then send" flow & save time in the long run is a "try and recover" flow, with something like this:

```ts
public async sendEvent(roomId: string, content: any): Promise<string> {
    await this.ensureRegistered(roomId);
    try {
        return this.client.sendMessage(roomId, content);
    } catch (err) {
        if (errorIsMembershipError(err)) { // this function doesn't exist, but you get the idea
            await this.client.joinRoom(roomId);
            return this.client.sendMessage(roomId, content);
        } else {
            throw err;
        }
    }
}
```